kyo-core/shared/src/main/scala/kyo/Async.scala

```diff
@@ -20,25 +20,57 @@ import scala.util.control.NonFatal
   *
+  * Note: For collection operations, Async provides concurrent execution variants of the `Kyo` companion object sequential operations. These
+  * operations use a bounded concurrency model to prevent resource exhaustion, defaulting to twice the number of available processors. This
+  * can be overridden per operation when needed. On platforms like the JVM, these operations may execute in parallel using multiple threads.
+  * On single-threaded platforms like JavaScript, operations will be interleaved concurrently but not execute in parallel.
+  *
   * This effect includes IO in its effect set to handle both async and sync execution in a single effect.
   *
   * @see
-  *   [[Async.run]] for running asynchronous computations
+  *   [[Async.run]] for executing asynchronous computations and obtaining a Fiber
   * @see
-  *   [[Async.Fiber]] for managing asynchronous tasks
+  *   [[Fiber]] for the low-level primitive that powers async execution (primarily for library authors)
   * @see
-  *   [[Async.Promise]] for creating and managing promises
+  *   [[Async.race]] for racing multiple computations and getting the first result
   * @see
-  *   [[Async.parallel]] for running computations in parallel
+  *   [[Async.gather]] for concurrently executing multiple computations and collecting their sucessful results
   * @see
-  *   [[Async.race]] for racing multiple computations
+  *   [[Async.foreach]], [[Async.collect]], and their variants for concurrent collection processing with bounded concurrency
   */
 opaque type Async <: (IO & Async.Join) = Async.Join & IO
 
 object Async:
 
     sealed trait Join extends ArrowEffect[IOPromise[?, *], Result[Nothing, *]]
 
+    /** Default concurrency level for collection operations.
+      *
+      * This value determines the maximum number of concurrent operations that can run in parallel for collection processing methods like
+      * foreach, collect, and their variants. It defaults to twice the number of available processors.
+      *
+      * This default can be overridden in two ways:
+      *   1. Per operation by passing an explicit concurrency parameter
+      *   2. Globally by setting the "kyo.async.concurrency.default" system property
+      *
+      * Example of setting the system property:
+      * ```
+      * java -Dkyo.async.concurrency.default=4 MyApp
+      * ```
+      *
+      * Consider adjusting this based on:
+      *   - Nature of operations (CPU vs IO bound)
+      *   - Available system resources
+      *   - Specific performance requirements
+      *
+      * Note: This only affects collection processing methods. Operations like race and gather run with unbounded concurrency.
+      */
+    val defaultConcurrency =
+        import AllowUnsafe.embrace.danger
+        given Frame = Frame.internal
+        IO.Unsafe.evalOrThrow(System.property[Int]("kyo.async.concurrency.default", Runtime.getRuntime().availableProcessors() * 2))
+    end defaultConcurrency
```

Comment:
I'm following a more explicit approach where the user has to override the behavior in each method call otherwise the execution falls back to `cores * 2`. The most problematic cases I've seen in production in similar scenarios are collections with hundreds of elements so this bound should be enough to prevent issues.

I'm not 100% happy with the approach but it seems a good improvement that we can iterate on later.